## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("Resources/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
application_df.NAME.value_counts()

PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: NAME, Length: 19568, dtype: int64

In [3]:
application_df1 = application_df.apply(lambda x: x.mask(x.map(x.value_counts())<100, 'other') if x.name=='NAME' else x)
application_df1.NAME.value_counts()

other                                                                 25987
PARENT BOOSTER USA INC                                                 1260
TOPS CLUB INC                                                           765
UNITED STATES BOWLING CONGRESS INC                                      700
WASHINGTON STATE UNIVERSITY                                             492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                         408
PTA TEXAS CONGRESS                                                      368
SOROPTIMIST INTERNATIONAL OF THE AMERICAS INC                           331
ALPHA PHI SIGMA                                                         313
TOASTMASTERS INTERNATIONAL                                              293
MOST WORSHIPFUL STRINGER FREE AND ACCEPTED MASONS                       287
LITTLE LEAGUE BASEBALL INC                                              277
INTERNATIONAL ASSOCIATION OF LIONS CLUBS                                266
MOMS CLUB   

In [4]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df2 = application_df1.drop(["EIN"], axis='columns')
application_df2.head()

NAME APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0  other              T10       Independent          C1000    ProductDev   
1  other               T3       Independent          C2000  Preservation   
2  other               T5  CompanySponsored          C3000    ProductDev   
3  other               T3  CompanySponsored          C2000  Preservation   
4  other               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [5]:
# Determine the number of unique values in each column.
application_df2.nunique()

NAME                        31
APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [6]:
# Look at APPLICATION_TYPE value counts for binning
application_df2.APPLICATION_TYPE.value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [7]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = ['T17', 'T15', 'T29','T14', 'T25', 'T2', 'T12', 'T13']

# Replace in dataframe
for app in application_types_to_replace:
    application_df2['APPLICATION_TYPE'] = application_df2['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df2['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
T9         156
Other      120
Name: APPLICATION_TYPE, dtype: int64

In [8]:
# Look at CLASSIFICATION value counts for binning
clasval = application_df2.CLASSIFICATION.value_counts()
clasval

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [9]:
# You may find it helpful to look at CLASSIFICATION value counts >1
clasval2 = clasval[clasval >1]
clasval2

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [10]:
clasval3 = clasval2[clasval2 <100]
clasval3

C2800    95
C7100    75
C1300    58
C1280    50
C1230    36
C1400    34
C7200    32
C2300    32
C1240    30
C8000    20
C7120    18
C1500    16
C1800    15
C6000    15
C1250    14
C8200    11
C1238    10
C1278    10
C1235     9
C1237     9
C7210     7
C2400     6
C1720     6
C4100     6
C1257     5
C1600     5
C1260     3
C2710     3
C0        3
C3200     2
C1234     2
C1246     2
C1267     2
C1256     2
Name: CLASSIFICATION, dtype: int64

In [11]:
application_df3 = application_df2.apply(lambda x: x.mask(x.map(x.value_counts())<100, 'other') if x.name=='CLASSIFICATION' else x)
application_df3.CLASSIFICATION.value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
other      669
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: CLASSIFICATION, dtype: int64

In [12]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df4 = pd.get_dummies(application_df3)
application_df4.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  NAME_ALPHA PHI SIGMA  \
0       1     5000              1                     0   
1       1   108590              1                     0   
2       1     5000              0                     0   
3       1     6692              1                     0   
4       1   142590              1                     0   

   NAME_AMATEUR ATHLETIC UNION OF THE UNITED STATES INC  \
0                                                  0      
1                                                  0      
2                                                  0      
3                                                  0      
4                                                  0      

   NAME_AMERICAN ASSOCIATION OF UNIVERSITY WOMEN  NAME_CIVITAN INTERNATIONAL  \
0                                              0                           0   
1                                              0                           0   
2                                              0                           0   
3                                              0                           0   
4                                              0                           0   

   NAME_DEMOLAY INTERNATIONAL  \
0                           0   
1                           0   
2                           0   
3                           0   
4                           0   

   NAME_FARMERS EDUCATIONAL AND COOPERATIVE UNION OF AMERICA  \
0                                                  0           
1                                                  0           
2                                                  0           
3                                                  0           
4                                                  0           

   NAME_HABITAT FOR HUMANITY INTERNATIONAL INC  ...  INCOME_AMT_1-9999  \
0                                            0  ...                  0   
1                                            0  ...                  1   
2                                            0  ...                  0   
3                                            0  ...                  0   
4                                            0  ...                  0   

   INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                       0                         0                   0   
1                       0                         0                   0   
2                       0                         0                   0   
3                       1                         0                   0   
4                       0                         1                   0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0                 0                       0                0   
1                 0                       0                0   
2                 0                       0                0   
3                 0                       0                0   
4                 0                       0                0   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0                  0                         1                         0  
1                  0                         1                         0  
2                  0                         1                         0  
3                  0                         1                         0  
4                  0                         1                         0  

[5 rows x 82 columns]

In [13]:
# Split our preprocessed data into our features and target arrays
app_X = application_df4.drop(columns='IS_SUCCESSFUL')
app_y = application_df4['IS_SUCCESSFUL']

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(app_X, app_y, random_state=42)

In [14]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [20]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn = tf.keras.models.Sequential()

nn.add(tf.keras.layers.Dense(units=25, activation="relu", input_dim=81))

nn.add(tf.keras.layers.Dense(units=12, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))
# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_5 (Dense)             (None, 25)                2050      
                                                                 
 dense_6 (Dense)             (None, 12)                312       
                                                                 
 dense_7 (Dense)             (None, 1)                 13        
                                                                 
Total params: 2,375
Trainable params: 2,375
Non-trainable params: 0
_________________________________________________________________


In [21]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [22]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5361 - accuracy: 0.7259
Epoch 2/100
804/804 [==============================] - 1s 2ms/step - loss: 0.4969 - accuracy: 0.7527
Epoch 3/100
804/804 [==============================] - 1s 2ms/step - loss: 0.4920 - accuracy: 0.7551
Epoch 4/100
804/804 [==============================] - 1s 2ms/step - loss: 0.4897 - accuracy: 0.7550
Epoch 5/100
804/804 [==============================] - 1s 2ms/step - loss: 0.4878 - accuracy: 0.7574
Epoch 6/100
804/804 [==============================] - 1s 2ms/step - loss: 0.4865 - accuracy: 0.7583
Epoch 7/100
804/804 [==============================] - 1s 2ms/step - loss: 0.4850 - accuracy: 0.7600
Epoch 8/100
804/804 [==============================] - 1s 2ms/step - loss: 0.4850 - accuracy: 0.7596
Epoch 9/100
804/804 [==============================] - 1s 2ms/step - loss: 0.4838 - accuracy: 0.7597
Epoch 10/100
804/804 [==============================] - 1s 2ms/step - loss: 0.4837 - accura

In [23]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.4972 - accuracy: 0.7513 - 451ms/epoch - 2ms/step
Loss: 0.49716249108314514, Accuracy: 0.7512536644935608


In [19]:
# Export our model to HDF5 file
nn.save('model.h5')

# Step 3: Optimize the Model

Put back in the NAME column.
Added more neurons to model.

# Step 4: Write a Report on the Neural Network Model

## Overview of the analysis: 
Explain the purpose of this analysis.

The purpose of this analysis is to have a model that can determine, with relative accuracy, whether or not an application will be successful after receiving funding based on certain characteristics.

## Results: 
Using bulleted lists and images to support your answers, address the following questions.

### Data Preprocessing

What variable(s) are considered the target(s) for your model?
What variable(s) are considered to be the features for your model?
What variable(s) are neither targets nor features, and should be removed from the input data?

The target of the model was the IS_SUCCESSFUL column as that is what is really important to the applicants.

The NAME was very helpful in determining success.

The EIN column was not useful to the model.

### Compiling, Training, and Evaluating the Model

How many neurons, layers, and activation functions did you select for your neural network model, and why?
Were you able to achieve the target model performance?
What steps did you take to try and increase model performance?

I added back in the NAME column to the data as well as adding more neurons and a couple layers to the model. The increase in information gained from the name column allowed the model performance to reach 75.1% on the test set. This is pretty good and an improvement from the earlier attempts.

## Summary: 
Summarize the overall results of the deep learning model. Include a recommendation for how a different model could solve this classification problem, and explain your recommendation.



Overall this model did a good job of training and testing given the data. Another model that could have been used here would have been KNN or a Random Forest model as those are both good at classification problems.